fem/fespace.cpp

```diff
@@ -93,16 +92,68 @@ FiniteElementSpace::FiniteElementSpace(const FiniteElementSpace &orig,
    Constructor(mesh, NURBSext, fec, orig.vdim, orig.ordering);
 }
 
-int FiniteElementSpace::GetOrder(int i) const
+void FiniteElementSpace::SetElementOrder(int i, int p)
 {
-   Geometry::Type GeomType = mesh->GetElementBaseGeometry(i);
-   return fec->FiniteElementForGeometry(GeomType)->GetOrder();
+   MFEM_VERIFY(sequence == mesh->GetSequence(), "space has not been Updated()");
+   MFEM_VERIFY(i >= 0 && i < GetNE(), "invalid element index");
+   MFEM_VERIFY(p >= 0 && p <= MaxVarOrder, "order ouf of range");
+   MFEM_ASSERT(!elem_order.Size() || elem_order.Size() == GetNE(),
+               "internal error");
+
+   if (elem_order.Size()) // already a variable-order space
+   {
+      if (elem_order[i] != p)
+      {
+         elem_order[i] = p;
+         orders_changed = true;
+      }
+   }
+   else // convert space to variable-order space
+   {
+      elem_order.SetSize(GetNE());
+      elem_order = fec->GetOrder();
+
+      elem_order[i] = p;
+      orders_changed = true;
+   }
+}
+
+int FiniteElementSpace::GetElementOrder(int i) const
+{
+   MFEM_VERIFY(sequence == mesh->GetSequence(), "space has not been Updated()");
+   MFEM_VERIFY(i >= 0 && i < GetNE(), "invalid element index");
+   MFEM_ASSERT(!elem_order.Size() || elem_order.Size() == GetNE(),
+               "internal error");
+
+   return GetElementOrderImpl(i);
 }
 
-int FiniteElementSpace::GetFaceOrder(int i) const
+int FiniteElementSpace::GetElementOrderImpl(int i) const
 {
-   Geometry::Type GeomType = mesh->GetFaceBaseGeometry(i);
-   return fec->FiniteElementForGeometry(GeomType)->GetOrder();
+   // (this is an internal version of GetElementOrder without asserts and checks)
+   return elem_order.Size() ? elem_order[i] : fec->GetOrder();
+}
+
+int FiniteElementSpace::GetEdgeOrder(int edge, int variant) const
+{
+   if (!IsVariableOrder()) { return fec->GetOrder(); }
+
+   const int* beg = var_edge_dofs.GetRow(edge);
+   const int* end = var_edge_dofs.GetRow(edge + 1);
+   if (variant >= end - beg) { return -1; } // past last variant
```

Comment:
Wouldn't it make more sense to throw an error?

Reply:
This is for consistency with `GetEdgeDofs` where the -1 is returned at the end when looping over all variants. Here it's strictly not necessary, but since we have no method to return the total number of variants for an edge/face this is "correct" return value at the moment. One may ask if looping until you get -1 is the best design, but that's a different question.